cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/SearchRequest.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package co.cask.cdap.data2.metadata.dataset;
+
+import co.cask.cdap.proto.EntityScope;
+import co.cask.cdap.proto.element.EntityTypeSimpleName;
+import co.cask.cdap.proto.id.NamespaceId;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A Metadata search request.
+ */
+public class SearchRequest {
+  private final NamespaceId namespace;
+  private final String query;
+  private final Set<EntityTypeSimpleName> types;
+  private final SortInfo sortInfo;
+  private final int offset;
+  private final int limit;
+  private final int numCursors;
+  @Nullable
+  private final String cursor;
+  private final boolean showHidden;
+  private final Set<EntityScope> entityScope;
+
+  public SearchRequest(NamespaceId namespace, String query, Set<EntityTypeSimpleName> types, SortInfo sortInfo,
+                       int offset, int limit, int numCursors, @Nullable String cursor, boolean showHidden,
+                       Set<EntityScope> entityScope) {
+    if (offset < 0) {
```

Comment:
not a part of this PR: but the numCursors can't be negative either right?

Reply:
I don't think so. will add a check.